test/uk/gov/hmrc/apiplatformmicroservice/thirdpartyapplication/controllers/ApplicationUpdateControllerSpec.scala

```diff
@@ -0,0 +1,139 @@
+import uk.gov.hmrc.apiplatformmicroservice.thirdpartyapplication.mocks._
+import uk.gov.hmrc.apiplatformmicroservice.thirdpartyapplication.services.UpliftApplicationService
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.Future.successful
+
+class ApplicationUpdateControllerSpec extends AsyncHmrcSpec with GuiceOneAppPerSuite with ApiDefinitionTestDataHelper {
+
+  trait Setup extends ApplicationByIdFetcherModule with ApplicationUpdateServiceModule with ApplicationBuilder with CollaboratorsBuilder with ApplicationJsonFormatters{
+    implicit val headerCarrier = HeaderCarrier()
+    implicit val mat = app.materializer
+
+    val applicationId = ApplicationId.random
+
+    val mockAuthConfig = mock[AuthConnector.Config]
+    val mockAuthConnector = mock[AuthConnector]
+
+
+    val controller = new ApplicationUpdateController(
+      mockAuthConfig,
+      mockAuthConnector,
+      ApplicationByIdFetcherMock.aMock,
+      ApplicationUpdateServiceMock.aMock,
+      Helpers.stubControllerComponents()
+    )
+  }
+
+  "addCollaborator" should {
+    val payload =
+      s"""{
+         |        "actor": {
+         |          "email": "someone@digital.hmrc.gov.uk",
+         |          "actorType": "COLLABORATOR"
+         |        }
+         |        ,
+         |        "collaboratorEmail": "thecollaborator@digital.hmrc.gov.uk"
+         |        ,
+         |        "collaboratorRole": "ADMINISTRATOR"
+         |        ,
+         |        "timestamp": "2022-10-12T08:06:46.706"
+         |        ,
+         |        "updateType": "addCollaboratorRequest"
+         |      }""".stripMargin
+
+
+    "return Ok when update service is successful" in new Setup {
+      val application = buildApplication(appId = applicationId)
+      val collaborator = buildCollaborator("bob@example.com", Role.DEVELOPER)
+      ApplicationByIdFetcherMock.FetchApplication.willReturnApplication(Option(application))
+      val request = FakeRequest("PATCH", s"/applications/${applicationId.value}")
+
+      ApplicationUpdateServiceMock.UpdateApplication.willReturnApplication(application)
+
+      val result = controller.update(applicationId)(request.withBody(Json.parse(payload)))
+
+      status(result) shouldBe OK
+    }
+
+    "return Not Found when no application is found" in new Setup {
+      val application = buildApplication(appId = applicationId)
+      val collaborator = buildCollaborator("bob@example.com", Role.DEVELOPER)
+      ApplicationByIdFetcherMock.FetchApplication.willReturnApplication(None)
+      val request = FakeRequest("PATCH", s"/applications/${applicationId.value}")
+
+      ApplicationUpdateServiceMock.UpdateApplication.willReturnApplication(application)
+
+      val result = controller.update(applicationId)(request.withBody(Json.parse(payload)))
+
+      status(result) shouldBe NOT_FOUND
+    }
+
+  }
+
+
+  "removeCollaborator" should {
```

Comment:
Do we need to test each event type in this spec? There is only one method under test (`update`) and the test is mocking the type-dependent specifics. At best, we're testing the parsing of the payload, but that could be done once with a bad payload for one of the tests to check that it returns 400? Otherwise this spec could get quite long as we add a test for each event type.

Reply:
You are correct, adds little. Removed